feat(orderbook/swaps): release order hold

sangaman opened the pull request:
This releases the hold placed on an order when we accept a swap once the swap either completes or fails. Previously, holds would remain on orders post-swap which would make part or all of the order unavailable for future swaps.

Closes #549.

moshababo commented on lib/orderbook/OrderBook.ts:
```diff
@@ -98,7 +98,14 @@ class OrderBook extends EventEmitter {
+          // if our order is the maker and the swap failed after it was agreed to but before it was executed
+          // we must release the hold on the order that we set when we agreed to the deal
+          const order = this.getOwnOrder(deal.orderId, deal.pairId);
+          order.hold! -= deal.quantity!;
```
this breaks the repository encapsulation.

sangaman replied:
I don't get what you mean, could you elaborate? Do you mean this line should be in TradingPair.ts?

moshababo replied:
Yes, I means repository in the greater sense, not our db repository class. `TradingPair` now manages the orders in-memory, so all the CRUD operations should be done through its API.

sangaman replied:
Alright that makes sense, thanks for clarifying.

moshababo commented on lib/orderbook/OrderBook.ts:
```diff
@@ -434,6 +441,12 @@ class OrderBook extends EventEmitter {
+      if (takerPubKey && !removeResult.fullyRemoved) {
+        // if takerPubKey is specified, it means we are removing quantity from the order due to a successful swap
+        // we must also reduce the amount that was put on hold while the swap was pending for the remaining order
+        removeResult.order.hold! -= quantityToRemove!;
```
same here

moshababo replied:
should use `tp.removeOrderHold`

sangaman commented:
@moshababo I put all the logic that updates the hold amount in TradingPair.ts. it feels a little awkward vs leaving the logic in OrderBook.ts, but if it looks good to you I'll combine the commits and merge.

moshababo commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,16 @@ class TradingPair {
     return maps.buy.get(orderId) || maps.sell.get(orderId);
   }
 
+  public addOrderHold = (orderOrOrderId: string | StampedOwnOrder, holdAmount: number) => {
```
it's indeed don't make much sense to have this method accept an order instance, and perform the same logic as previously `OrderBook` did. I think it should only accept `orderId`, even though you have the order instance sometimes. Otherwise it's still breaking the encapsulation, and reducing `addOrderHold` to a util function. 
The point is not to encourage updates via instance mutations, although they work at the moment. You might get an instance copy for example, and so it won't work, or that other actions need to be applied in the repository (like in `order.quantity = 0`).

moshababo commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,16 @@ class TradingPair {
 
+  public addOrderHold = (orderOrOrderId: string | StampedOwnOrder, holdAmount: number) => {
+    const order = typeof orderOrOrderId === 'string' ? this.getOwnOrder(orderOrOrderId) : orderOrOrderId;
+    order.hold = order.hold ? order.hold + holdAmount : holdAmount;
```
what about if `hold` exceeds `quantity`?
you're relying on higher-layer code to protect against that. I'd suggest to encapsulate that in what you're doing here, since it's not a generic update method, but rather specifically for `hold`.

sangaman commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,16 @@ class TradingPair {
+
+  public removeOrderHold = (orderOrOrderId: string | StampedOwnOrder, holdAmount: number) => {
+    const order = typeof orderOrOrderId === 'string' ? this.getOwnOrder(orderOrOrderId) : orderOrOrderId;
+    order.hold = order.hold ? order.hold + holdAmount : holdAmount;
```
Thanks for the catch, copied the method and forgot to change the operator.

sangaman commented:
@moshababo Thanks for the review. I addressed your feedback - one thing I left as is was having separate add and remove order hold methods. It felt a little clumsy to have to multiply parameters by negative one, plus I realized some `assert` statements were in order which made the two methods more distinct from each other. I also went ahead and added unit tests for these methods to make sure the logic was correct.

erkarl commented on test/unit/TradingPair.spec.ts:
```diff
@@ -339,3 +339,57 @@ describe('MatchingEngine queues and maps integrity', () => {
   });
 });
+
+describe('TradingPair.addOrderHold & TradingPair.removeOrderHold', () => {
```
Really like this addition...

moshababo commented on lib/orderbook/OrderBook.ts:
```diff
@@ -98,7 +98,14 @@ class OrderBook extends EventEmitter {
+        if (deal.role === SwapRole.Maker && deal.phase === SwapPhase.SwapAgreed) {
+          // if our order is the maker and the swap failed after it was agreed to but before it was executed
+          // we must release the hold on the order that we set when we agreed to the deal
+          const tp = this.tradingPairs.get(deal.pairId);
```
can use `this.getTradingPair` as well, since you'll get an error anyway if it's missing (but less informative one)

moshababo commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,18 @@ class TradingPair {
+  public addOrderHold = (orderId: string, holdAmount: number) => {
+    const order = this.getOwnOrder(orderId);
+    assert((order.hold || 0) + holdAmount <= order.quantity, 'the amount of an order on hold cannot exceed the available quantity');
+    order.hold = order.hold ? order.hold + holdAmount : holdAmount;
```
nothing serious, but you're checking twice if `order.hold` exists, and initializing it anyway. Instead you can just initialize it (`order.hold = order.hold || 0`) in the beginning, and avoid the checks.

sangaman replied:
@offerm suggested [here](https://github.com/ExchangeUnion/xud/pull/627#discussion_r231400551) to make hold initialize to 0 to simplify this logic, which I think is a fine idea, so I'll just go with that approach if that's alright with you.

moshababo replied:
Sure.

offerm commented on lib/orderbook/OrderBook.ts:
```diff
@@ -434,6 +442,12 @@ class OrderBook extends EventEmitter {
         this.pool.broadcastOrderInvalidation(removeResult.order, takerPubKey);
       }
 
+      if (takerPubKey && !removeResult.fullyRemoved) {
```
See comment above.
the condition on takerPubKey looks not in place. 
Also, why not to clear to hold amount even if the order is fully removed? Yes, it is redundant but the code will be much clearer.

I also  see no reason why not to hold/release for local matched.

offerm commented on lib/orderbook/OrderBook.ts:
```diff
@@ -434,6 +442,12 @@ class OrderBook extends EventEmitter {
         this.pool.broadcastOrderInvalidation(removeResult.order, takerPubKey);
```
Question: Are we sending invalidation even if !removeResult.fullyRemoved?

offerm replied:
I don't understand. why to send invalidation of the order is still valid (after partial swap)?

sangaman replied:
It lets peers know that part of the order is no longer available, so they know to reduce the quantity on their end.

offerm replied:
not the best name for the operation.
As a minimum,  I think we should add a comment explaining what this method does.

In addition, while a P2P message was sent to multiple nodes, there is nothing in the logs (senders and receivers) which indicates that a message was sent accepted and processed (not to mention the outcome. Suggest you add this as part of #649 (or a new issue if you like).

erkarl replied:
Good catch @offerm.

It's definitely confusing as of now.

sangaman replied:
I'll add comments for the packet in a separate PR, but we do have that logging [when the packet is received and processed](https://github.com/ExchangeUnion/xud/blob/3868681741bada5ac39166152f14530ee3873f3a/lib/p2p/Pool.ts#L490) as well as logging when an [order is removed or reduced for any reason](https://github.com/ExchangeUnion/xud/blob/3868681741bada5ac39166152f14530ee3873f3a/lib/orderbook/TradingPair.ts#L206). Let's continue this in #649.

kilrau replied:
Need to use verbose mode @offerm

sangaman replied:
Verbose messages should be logged by default, since the default level is `debug` which logs everything except `trace`.

offerm commented on lib/orderbook/OrderBook.ts:
```diff
@@ -540,7 +554,8 @@ class OrderBook extends EventEmitter {
         // quantityToAccept = Math.min(proposedQuantity, availableQuantity);
         const quantityToAccept = proposedQuantity;
```
Please remove order.hold ? from the definition of availableQuantity

offerm replied:
I would add an assert to make sure that availableQuantity is positive

offerm commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,18 @@ class TradingPair {
   }
 
+  public addOrderHold = (orderId: string, holdAmount: number) => {
+    const order = this.getOwnOrder(orderId);
```
assert holdamount is positive

offerm commented on lib/orderbook/TradingPair.ts:
```diff
@@ -267,6 +267,18 @@ class TradingPair {
+  }
+
+  public removeOrderHold = (orderId: string, holdAmount: number) => {
+    const order = this.getOwnOrder(orderId);
```
assert holdamount is positive

offerm commented:
Looks good. Some small changes please

sangaman commented on lib/orderbook/OrderBook.ts:
```diff
@@ -98,7 +98,14 @@ class OrderBook extends EventEmitter {
           this.emit('ownOrder.swapped', { pairId, quantity, id: orderId });
```
Yeah that does seem like a more sensible flow, I'm changing it.

sangaman commented:
Thanks for the review @offerm, I believe I addressed all your comments.
